plugins/database/quota/src/org/apache/cloudstack/quota/QuotaServiceImpl.java

```
@@ -0,0 +1,304 @@
+            List<QuotaBalanceVO> qbrecords = _quotaBalanceDao.lastQuotaBalanceVO(accountId, domainId, adjustedStartDate);
+            if (s_logger.isDebugEnabled()){
+                s_logger.debug("Found records size=" + qbrecords.size());
+            }
+            if (qbrecords.size() == 0) {
+                throw new InvalidParameterValueException("Incorrect Date there are no quota records before this date " + adjustedStartDate);
+            } else {
+                return qbrecords;
+            }
+        } else {
+            Date adjustedStartDate = computeAdjustedTime(startDate);
+            if (endDate.after(_respBldr.startOfNextDay())) {
+                throw new InvalidParameterValueException("Incorrect Date Range. End date:" + endDate + " should not be in future. ");
+            } else if (startDate.before(endDate)) {
+                Date adjustedEndDate = computeAdjustedTime(endDate);
+                if (s_logger.isDebugEnabled()){
+                    s_logger.debug("getQuotaBalance2: Getting quota balance records for account: " + accountId + ", domainId: " + domainId + ", between " + adjustedStartDate + " and " + adjustedEndDate);
+                }
+                List<QuotaBalanceVO> qbrecords = _quotaBalanceDao.findQuotaBalance(accountId, domainId, adjustedStartDate, adjustedEndDate);
+                if (s_logger.isDebugEnabled()){
+                    s_logger.debug("getQuotaBalance3: Found records size=" + qbrecords.size());
+                }
+                if (qbrecords.size() == 0) {
+                    throw new InvalidParameterValueException("Incorrect Date range there are no quota records between these dates start date " + adjustedStartDate + " and end date:" + endDate);
+                } else {
+                    return qbrecords;
+                }
+            } else {
+                throw new InvalidParameterValueException("Incorrect Date Range. Start date: " + startDate + " is after end date:" + endDate);
```

Why not check the parameter values at the top of the method to fail early and reduce the nesting of the ``if`` blocks.